curvefs/src/metaserver/metastore.cpp

```diff
@@ -0,0 +1,521 @@
+    auto callbackFunc = [&](ENTRY_TYPE type, uint32_t paritionId,
+                            void* metadata) -> bool {
+        if (type == ENTRY_TYPE::PARTITION) {
+            PartitionInfo* partition =
+                reinterpret_cast<PartitionInfo*>(metadata);
+            partitionMap_.emplace(partition->partitionid(),
+                                  std::make_shared<Partition>(*partition));
+            return true;
+        } else if (type == ENTRY_TYPE::INODE) {
+            std::shared_ptr<Partition> partition = GetPartiton(paritionId);
+            if (partition == nullptr) {
+                LOG(ERROR) << "MetaStore Load GetPartiton fail, type INODE";
+                return false;
+            }
+            Inode* inode = reinterpret_cast<Inode*>(metadata);
+            MetaStatusCode status = partition->InsertInode(*inode);
+            return (status == MetaStatusCode::OK) ? true : false;
+        } else if (type == ENTRY_TYPE::DENTRY) {
+            std::shared_ptr<Partition> partition = GetPartiton(paritionId);
+            if (partition == nullptr) {
+                LOG(ERROR) << "MetaStore Load GetPartiton fail, type DENTRY";
+                return false;
+            }
+            Dentry* dentry = reinterpret_cast<Dentry*>(metadata);
+            MetaStatusCode status = partition->CreateDentry(*dentry);
+            if (status != MetaStatusCode::OK) {
+                LOG(ERROR) << "MetaStore Load CreateDentry fail, type DENTRY";
+                return false;
+            }
+            return true;
+        } else {
+            LOG(ERROR) << "load meta from file, not supported type";
+            return false;
+        }
+        return true;
+    };
+
+    // Load from raft snap file to memory
+    WriteLockGuard writeLockGuard(rwLock_);
+    bool ret = LoadFromFile(path, callbackFunc);
+    if (!ret) {
+        partitionMap_.clear();
+    }
+    return ret;
+}
+
+void MetaStore::SaveBack(const std::string& path, OnSnapshotSaveDone* done) {
+    // 1. call back dump function
+    std::vector<std::shared_ptr<Iterator>> iteractorList;
+    LOG(INFO) << "Save meta to file back.";
+    // add partition iteractor
+    std::map<uint32_t, PartitionInfo> partitionInfoMap;
+    for (auto partitionIter : partitionMap_) {
+        partitionInfoMap.emplace(partitionIter.second->GetPartitionId(),
+                                 partitionIter.second->GetPartitionInfo());
+    }
+    // partitionIter no need pass partitionId, here pass 0 instead
+    std::shared_ptr<ContainerIterator<std::map<uint32_t, PartitionInfo>>>
+        partitionContainerIter = std::make_shared<
+            ContainerIterator<std::map<uint32_t, PartitionInfo>>>(
+            ENTRY_TYPE::PARTITION, 0, &partitionInfoMap);
+    iteractorList.push_back(partitionContainerIter);
+
+    // add inode and dentry iterator
+    for (auto it : partitionMap_) {
+        // add inode iterator
+        std::shared_ptr<ContainerIterator<std::unordered_map<
```

Comment:
The constructor is too complex, maybe it can simplify.

Reply:
done